databroker/broker.py

```diff
@@ -880,3 +881,152 @@ def _munge_time(t, timezone):
+    As a side effect, documents are inserted into the databases and external
+    files may be written.
+
+    Parameters
+    ----------
+    db: ``databroker.Broker`` instance
+        The databroker to store the documents in, must have writeable
+        metadatastore and writeable filestore if ``external`` is not empty.
+    external_writers : dict
+        Maps data keys to a ``WriterClass``, which is responsible for writing
+        data to disk and creating a record in filestore. It will be
+        instantiated (possible multiple times) with the argument ``db.fs``.
+        If it requires additional arguments, use ``functools.partial`` to
+        produce a callable that requires only ``db.fs`` to instantiate the
+        ``WriterClass``.
+        """
+    if external_writers is None:
+        external_writers = {}  # {'name': WriterClass}
+
+    def wrap(f):
+        def wrapped_f(*args, **kwargs):
+            gen = f(*args, **kwargs)
+            for name, doc in gen:
+                # doc will pass through unchanged; fs_doc may be modified to
+                # replace some values with references to filestore.
+                if external_writers:
+                    fs_doc = copy.deepcopy(doc)
+                else:
+                    fs_doc = doc  # for perf
+
+                if name == 'start':
+                    # Make a fresh instance of any WriterClass classes.
+                    writers = {data_key: cl(db.fs)
+                               for data_key, cl in external_writers.items()}
+
+                if name == 'descriptor':
+                    # Mutate fs_doc here to mark data as external.
+                    for data_name in external_writers.keys():
+                        fs_doc['data_keys'][data_name].update(
+                            external='FILESTORE:')
+
+                elif name == 'event':
+                    # The writer writes data to an external file, creates a
+                    # datum record in the filestore database, and return that
+                    # datum_id. We modify fs_doc in place, replacing the data
+                    # values with that datum_id.
+                    for data_key, writer in writers.items():
+                        fs_uid = writer.write(fs_doc['data'][data_key])
+                        fs_doc['data'][data_key] = fs_uid
+
+                    doc.update(
+                        filled={k: True for k in external_writers.keys()})
+
+                elif name == 'stop':
+                    for data_key, writer in list(writers.items()):
+                        writer.close()
+                        writers.pop(data_key)
+
+                # The (maybe mutated) fs_doc is inserted into metadatastore.
+                db.mds.insert(name, fs_doc)
+
+                # The pristine doc is yielded.
+                yield name, doc
+
+        return wrapped_f
+
+    return wrap
+
+
+def event_map(stream_name, data_keys, provenance):
```

Comment:
Yes, likely some of it will be automatically captured: at least we can know the name of the function and that it was applied to events by `event_map`. Even though I think we have a good of some things that will be in provenance, a long discussion is needed to come up with a complete plan for its contents. I'm completely punting for now.